src/display.rs

```diff
@@ -538,6 +538,43 @@ where
     f.pad_integral(true, "0x", encoded)
 }
 
+/// Given a `T:` [`fmt::Write`], `HexWriter` implements [`std::io::Write`]
+/// and writes the source bytes to its inner `T` as hex characters.
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+pub struct HexWriter<T>(T, Case);
+
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+impl<T> HexWriter<T> {
+    /// Creates a `HexWriter` that writes the source bytes to `dest` as hex characters
+    /// in the given `case`.
+    pub fn new(dest: T, case: Case) -> Self { Self(dest, case) }
+    /// Consumes this `HexWriter` returning the inner `T`.
+    pub fn into_inner(self) -> T { self.0 }
+}
+
+#[cfg(any(test, feature = "std"))]
+#[cfg_attr(docsrs, doc(cfg(any(test, feature = "std"))))]
+impl<T> std::io::Write for HexWriter<T>
+where
+    T: core::fmt::Write,
+{
+    fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+        let mut n = 0;
+        let table = self.1.table();
+        for byte in buf {
+            let pair = table.byte_to_hex(*byte);
+            if self.0.write_str(&pair).is_err() {
+                break;
```

Comment:
This needs to return `Err(std::io::ErrorKind::Other.into())`.

Reply:
Thank you for the review. Help me understand:
- In [`std::io::Write::write`](https://doc.rust-lang.org/stable/std/io/trait.Write.html#tymethod.write) docs, they say _"If an error is returned then no bytes in the buffer were written to this writer."_ So here we would need to somehow "clear" the `T` upon returning an `std::io::Error`? Not sure of the correct approach to do this given `T: fmt::Write` alone.
- As an example, when `ArrayString::write_str` returns an `fmt::Error`, it is only an `arrayvec::CapacityError`. It seems to me this matches better to an `Ok(0)` return value from `std::io::Write` rather than a `std::io::Error` because _"It is not considered an error if the entire buffer could not be written to this writer."_ (from `std::io::Write` docs)
- Then there is this piece from [`core::fmt::Error`](https://doc.rust-lang.org/stable/core/fmt/struct.Error.html): _"This type does not support transmission of an error other than that an error occurred. This is because, despite the existence of this error, string formatting is considered an infallible operation."_

I can see how propagating `ErrorKind::Other` could be a nice way to tell people _"hey we don't know what happened, but `fmt::Write::write_str` returned an error."_ If we go with this approach, I would need to figure out a way to clear the "T" as described above.

Reply:
Sadly, I think we can't follow the contract exactly but it should still work fine for consumers not relying on it (not doing any async operations). We should certainly document that it's imprecise.

I was originally thinking of also having `io::Write` -> `io::Write` but discarded the idea because I thought people could just use `fmt2io`. I was just thinking of adding it anyway but there's still an edge case when the underlying writer writes only one hex digit - we can't return 0.5.

Reply:
Maybe if `n > 0` we should just truncate (as is done) to indicate "partial write succeeded". But if `n == 0` (i.e. the user tried to continue a partial write or called `write_all` which'll do it for him) then we return the actual error, which we will hit immediately.

I think this upholds the io::Write contract without requiring any exotic tricks.

We should have comments clearly explaining what we're doing here.

Reply:
> But if n == 0 (i.e. the user tried to continue a partial write or called write_all which'll do it for him) then we return the actual error, which we will hit immediately.

I went with this solution in eaea55bfbe440fb659e64ef617ddec20a2010539 below - we return the error if n == 0, **and** the provided buffer is not empty.

Reply:
@apoelstra I guess it can work but it will have worse performance. We have to write digit-by-digit and buffer the unwritten second digit if any. Then report success on number of bytes and implement `flush()` to attempt writing the buffered byte.